File: sdk/python/rhoban/robot.py
#!/usr/bin/python
# -*- coding: utf-8 -*-

import sys, os, re, threading, time
import yaml
import rhoban.configurations as config
import rhoban.communication as com
import rhoban.motors as motors
import rhoban.motion as motion
import rhoban.sensors as sensors

"""
    Représente un groupe de robots
"""
class Robots(object):
    def __init__(self):
        self.robots = {}

    def loadYaml(self, filename):
        config = yaml.load(open(filename, 'rb').read())
        storeFileName = '../common/commands.xml'

        if 'commands' in config:
            storeFileName = config['commands']

        if 'robots' not in config or not config['robots']:
            raise Exception('Config error: no "robots" entry')

        for robotName, robotConfig in config['robots'].items():
            if not robotConfig.get('enabled', True):
                continue
            
            robot = Robot(storeFileName)
            self.robots[robotName] = robot
            robot.name = robotName
            
            if 'host' in robotConfig:
                robot.connect(robotConfig['host'], robotConfig.get('port', 12345))

            if 'environment' in robotConfig:
                robot.loadEnvironment(robotConfig['environment'])

            if 'loadMoves' in robotConfig:
                loadedMoves = robot.getLoadedMoves()

                for move in robotConfig['loadMoves']:
                    if move not in loadedMoves:
                        loadedMoves.append(move)
                        robot.loadMove(move)

    def stop(self):
        for name, robot in self.robots.items():
            robot.stop()

    def __len__(self):
        return len(self.robots)

    def __iter__(self):
        return iter(self.robots)

    def __getitem__(self, name):
        return self.robots[name]

"""
    Ensemble de fonctions et raccourcis pour communiquer avec un robot
"""
class Robot(object):
    def __init__(self, storeFileName = '../common/commands.xml'):
        self.store = com.CommandsStore()
        self.store.parseXml(storeFileName)

        self.connection = com.Connection()
        self.connection.setStore(self.store)

        self.moves = motion.Moves(self.connection)

        self.motors = motors.Motors(self.connection)
        self.sensors = sensors.Sensors(self.connection)
        self.configs = config.Configurations(self.connection)

        self.environment = ''

    # Environment

    def loadEnvironment(self, environment):
        self.setEnvironment(environment)
        if os.path.exists(os.path.join(environment, 'LowLevelConfig.xml')):
            self.loadLowLevelConfig(os.path.join(environment, 'LowLevelConfig.xml'))
        else:
            self.loadLowLevelConfig(os.path.join(environment, 'ConfigFiles', 'LowLevelConfig.xml'))

        if os.path.exists(os.path.join(environment, 'MoveSchedulerConfig.xml')):
            self.loadMoveSchedulerConfig(os.path.join(environment, 'MoveSchedulerConfig.xml'))
        else:
            self.loadMoveSchedulerConfig(os.path.join(environment, 'ConfigFiles', 'MoveSchedulerConfig.xml'))

    def setEnvironment(self, environment):
        self.environment = environment

    # Connection

    def connect(self, hostname = 'localhost', port = 12345):
        self.connection.connectTo(hostname, port)

    def isConnected(self):
        return self.connection.connected

    def serverVersion(self):
        response = self.connection.ServerGetVersion_response()
        return response[0]

    def testConnection(self):
        print('Testing server version...')
        response = self.connection.ServerGetVersion_response()
        print('~> %s' % response[0])

        print('Testing echo "Hello world"...')
        response = self.connection.ServerEcho_response(b'Hello world')
        print('~> %s' % response[0])

        if response[0] == 'Hello world':
            print("\n"+'Connection test passed')

    # Configuration

    def loadLowLevelConfig(self, config, force = False):
        self.configs.loadLowLevelConfig(config, force)

    def loadMoveSchedulerConfig(self, config, force = False):
        self.configs.loadMoveSchedulerConfig(config, force)
        self.motors.setConfig(self.configs.moveSchedulerConfig)

    # Motors

    def allCompliant(self):
        self.motors.allCompliant()

    # Moves

    def moveFileName(self, name):
        return os.path.join(self.environment, 'Moves', name + '.xml')

    def loadMove(self, name):
        filename = self.moveFileName(name + '.graphics')

        if os.path.exists(filename):
            self.moves.loadMove(filename)
        else:
            self.moves.loadMove(self.moveFileName(name))

    def startMove(self, name, duration = 0, smooth = 500):
        self.moves.startMove(name, duration, smooth)

    def pauseMove(self, name):
        self.moves.pauseMove(name)

    def stopMove(self, name, smooth = 500):
        self.moves.stopMove(name, smooth)

    def killMove(self, name):
        self.moves.killMove(name)

    def getLoadedMoves(self):
        return self.moves.getLoadedMoves()

    def updateConstant(self, moveName, constantName, value):
        self.moves.updateConstant(moveName, constantName, float(value))

    def emergency(self):
        self.connection.SchedulerEmergencyStop()
        self.motors.allCompliant()

    def stop(self):
        self.connection.stop()
        self.motors.stop()

File: sdk/cpp/rhoban/robot/Robot.cpp
/*************************************************
 * Publicly released by Rhoban System, August 2012
 *             www.rhoban-system.fr
 *
 * Freely usable for non-commercial purposes
 *
 * Licence Creative Commons *CC BY-NC-SA
 * http://creativecommons.org/licenses/by-nc-sa/3.0
 *************************************************/
#include <cstdlib>
#include <cstdio>
#include <iostream>
#include <string>
#include <vector>
#include <communication/Connection.h>
#include <communication/CommandsStore.h>
#include <communication/Message.h>
#include <types.h>
#include <motors/Motors.h>
#include <config/Configurations.h>
#include "Robot.h"
#include "Moves.h"
#include "Sensors.h"
#include "file_mngt.h"

using namespace std;

namespace Rhoban
{
Robot::Robot(CommandsStore *commandsStore, string name)
{
	connection = new Connection(commandsStore);
	motors = new Motors(connection);
	configs = new Configurations(connection);
	moves = new Moves(connection);
	sensors = new Sensors(connection);
   vision = new Vision(connection);
	this->name = name;
}

Robot::~Robot()
{
	delete configs;
	delete connection;
	delete motors;
	delete moves;
	delete sensors;
}

void Robot::loadEnvironment(string environment)
{
	try
	{
		this->setEnvironment(environment);
		checkFixEnvironmentPath();

                string env1, env2;

		env1 = this->environment + "LowLevelConfig.xml";

                if (file_exists(env1)) {
                    this->loadLowLevelConfig(env1);
                } else {
		    env1 = this->environment + "ConfigFiles/LowLevelConfig.xml";
        	    this->loadLowLevelConfig(env1);
                }
		
                env2 = this->environment + "MoveSchedulerConfig.xml";
                if (file_exists(env2)) {
    		    this->loadMoveSchedulerConfig(env2);
                } else {
                    env2 = this->environment + "ConfigFiles/MoveSchedulerConfig.xml";
		    this->loadMoveSchedulerConfig(env2);
                }
	}
	catch(string exc)
	{
		throw string("Failed to load environment ") + environment + " for robot " + name + " :\n\t" + exc;
	}
}

void Robot::checkFixEnvironmentPath()
{
	string::iterator it;
	it = environment.end();
	it--;
	if(*it != '/')
		environment.push_back('/');
}

void Robot::connect(const char *adress, int port)
{
	this->hostname = adress;
	this->port = port;
	connection->connectTo(adress, port);
}

int Robot::isConnected()
{
	return connection->isConnected();
}

ui32 Robot::serverVersion()
{
	try
	{
		Message response = connection->ServerGetVersion_response(10000);
		uint answer = response.read_uint();
		return answer;
	}
	catch(string exc)
	{
		throw string("Failed to get server version:\n\t") = exc;
	}
}

int Robot::testConnection()
{
	cout << "Testing server version..." << endl;

	cout << "Version : " << serverVersion() << endl;

	cout << "Testing echo\"Hello world\"..." << endl;

	Message response = connection->ServerEcho_response("Hello ","world", 10000);
	string s1 = response.read_string();
	string s2 = response.read_string();

	cout << " Echo : " << s1 << s2 << endl;

	if((s1 == "Hello ") && (s2 == "world"))
		cout << "Connection test successful." << endl;
	else
		cout << "Connection test failed." << endl;

}

void Robot::loadLowLevelConfig(string filename, bool force)
{
	configs->loadLowLevelConfig(filename, force);
}

void Robot::loadMoveSchedulerConfig(string filename, bool force)
{
	configs->loadMoveSchedulerConfig(filename, force);
	motors->setConfig(configs->getMoveSchedulerConfig());
}

void Robot::allCompliant()
{
	motors->allCompliant();
}

string Robot::moveFileName(string name)
{
	string retval = getEnvironment();
	retval.append("Moves/");
	retval.append(name);
	retval.append(".xml");
	return retval;
}

void Robot::loadMove(string name, bool rawPath)
{
	string filename = name;
        if (!rawPath) {
	    filename.append(".graphics");
	    filename = moveFileName(filename);
        }

	ifstream ifile(filename.c_str());
	bool fileExists = (bool)ifile;
	ifile.close();

	if(fileExists)
		moves->loadMove(filename);
	else
	{
		string filename2 = moveFileName(name);

		ifstream ifile2(filename2.c_str());
		bool fileExists2 = (bool)ifile2;
		ifile2.close();

		if(fileExists2)
			moves->loadMove(filename2);
		else
			cout << "Move file for " << name << " not found." << endl;
	}
}

void Robot::startMove(string name, ui32 duration, ui32 smooth)
{
	moves->startMove(name, duration, smooth);
}

void Robot::pauseMove(string name)
{
	moves->pauseMove(name);
}

void Robot::stopMove(string name, ui32 smooth)
{
	moves->stopMove(name, smooth);
}

void Robot::killMove(string name)
{
	moves->killMove(name);
}

vector<string> Robot::getLoadedMoves()
{
	return moves->getLoadedMoves();
}

void Robot::updateConstant(string moveName, string constantName, float value)
{
	moves->updateConstant(moveName, constantName, value);
}
    
void Robot::updateBlockXML(string moveName, string xml)
{
    moves->updateBlockXML(moveName, xml);
}

void Robot::emergency()
{
	connection->SchedulerEmergencyStop();
}

void Robot::stop()
{
	connection->stop();
	motors->stop();
}

void Robot::setMotors(Motors *motors)
{
	this->motors = motors;
}

Motors* Robot::getMotors()
{
	return motors;
}

void Robot::setConfigs(Configurations *configs)
{
	this->configs = configs;
}

Configurations *Robot::getConfigs()
{
	return configs;
}

void Robot::setConnection(Connection *connection)
{
	this->connection = connection;
}

Connection *Robot::getConnection()
{
	return connection;
}

void Robot::setEnvironment(string path)
{
	environment = path;
	checkFixEnvironmentPath();
}

string Robot::getEnvironment()
{
	checkFixEnvironmentPath();
	return environment;
}
void Robot::setMoves(Moves *moves)
{
	this->moves = moves;
}

Moves *Robot::getMoves()
{
	return moves;
}

void Robot::setSensors(Sensors *sensors)
{
	this->sensors = sensors;
}

Sensors *Robot::getSensors()
{
	return sensors;
}

void Robot::setVision(Vision *vision)
{
	this->vision = vision;
}

Vision *Robot::getVision()
{
	return vision;
}

void Robot::setName(string name)
{
	this->name = name;
}

string Robot::getName()
{
	return name;
}

void Robot::setHostname(string hostname)
{
	this->hostname = hostname;
}

string Robot::getHostname()
{
	return hostname;
}

void Robot::setPort(int port)
{
	this->port = port;
}

int Robot::getPort()
{
	return port;
}
}
